### PR TITLE
Implementation of r_set_select()

### PR DESCRIPTION
**Pull request by matthiasbeyer**
Targets #176

### CODE REVIEW

File: src/libreset/ll/ll.h [hunk 1]
```diff
@@ -167,6 +167,22 @@ __nonnull__(1)
...
+ * @memberof ll
+ *
+ * @return zero on success, else error code
+ */
+int
+ll_select(
+    struct ll const* src, //!< The source from where to select
+    r_predf pred, //!< The predicate
+    void* etc, //!< Additional information for the predicate function
+    int (*insf)(void* elem, void* insetc), //!< Function to insert somewhere
```
Comment (TheNeikos) [inline src/libreset/ll/ll.h]: `insetc` ?

**Comment by matthiasbeyer** [comment]
The naming is not very good by now. I'm to tired to fix this now, will do tomorrow!

File: src/libreset/ht/ht.h [hunk 2]
```diff
@@ -148,7 +164,7 @@ ht_ndel(
  */
 static inline size_t
 ht_nbuckets(
-    struct ht* ht //!< The hashtable object to calculate the bucket count for
+    struct ht const* const ht //!< The ht object to calc the bucket count for
```
Comment (TheNeikos) [inline src/libreset/ht/ht.h:203]: What is the connection to `r_set_select`?

File: src/libreset/ht/ht.h [hunk 3]
```diff
@@ -161,7 +177,7 @@ ht_nbuckets(
 
 static inline size_t
 ht_nbuckets(
-    struct ht* ht
+    struct ht const* const ht
```
Comment (TheNeikos) [inline src/libreset/ht/ht.h:218]: What is the connection to r_set_select?

Comment (matthiasbeyer) [inline src/libreset/ht/ht.h:218]: Nothing. Just found this bug. Will extract tomorrow and submit in a separate PR.

**Comment by neithernut** [comment]
The select function is supposed to put the selected elements into another set.

**Comment by matthiasbeyer** [comment]
@neithernut yes? It does...?

**Comment by neithernut** [comment]
I thought so...
Well, if we keep the vurrent definition, we should ensure that we can simply pass `set_insert()` as `insf`. This is currently not the case.

**Comment by matthiasbeyer** [comment]
This would require the linked list implementation to include the `libreset/set.h` header, which I consider as very ugly.

There for the helper function, which has a more generic interface.

**Comment by neithernut** [comment]
Nope, the function declaration just has to fit "coincidentally". (Which means that the function can be casted to the `insf` type)

**Comment by matthiasbeyer** [comment]
@neithernut I do not understand your comment.

**Comment by neithernut** [comment]
The parameters and return value has to be the same for `insf()` and `set_insert()` (except that `struct r_set*` is `void*` for `insf`).

**Comment by matthiasbeyer** [comment]
But why, we pass a function which does insertion. I do the insertion into the ht here, as the set is absolutely not relevant... so it is much simpler to insert into the ht of the destination directly.

**Comment by neithernut** [comment]
Yes, but if you want to give users the possibility to use a custom `insf()` to do _not_ use a set (like printing the selected elements directly), you should take care of the usecase of selecting elements into another set, too.
Also: the number of selected elements may potentially be quite low. In such cases, inserting one element after another does, indeed, make more sense.

**Comment by matthiasbeyer** [comment]
I guess we shouldn't provide such an possibility.

**Comment by neithernut** [comment]
Why not? This way we would need to implement the selection only once.
And the benefits of doing it twice are minimal in this case.

**Comment by matthiasbeyer** [comment]
@neithernut can you send some commits for this?

**Comment by neithernut** [comment]
Yep (after the switch to drone.io)

**Comment by matthiasbeyer** [comment]
I'm ready for merge. @neithernut , @TheNeikos SYN.

**Comment by neithernut** [comment]
Please rebase.

**Comment by neithernut** [comment]
And squash.

**Comment by neithernut** [comment]
ACK.
